### ISSUE TITLE
No way to run multiple Kernels in parallel on GPU and then wait until all are complete?

### ISSUE THREAD

**Issue by jonmdev** [ProjectPhysX/OpenCL-Wrapper#14]
## Background 

For background, regarding my prior problem I was describing [here](https://github.com/ProjectPhysX/OpenCL-Wrapper/issues/13), I did not realize we have in OpenCL [synchronization across a CUDA warp](https://stackoverflow.com/questions/10155499/opencl-synchronization) (ie. any Kernel run on NVIDIA with <=32 threads) with `barrier(CLK_LOCAL_MEM_FENCE);`

This solves the problem I was asking about there completely, as I can just run in OpenCL my full audio buffer calculation with one Kernel at max 32 threads per process that must stay synced and use the `barrier` to synchronize as needed throughout the calculation. 

So thanks yet again, as OpenCL is much easier to add to an existing project than CUDA and I will definitely be assisted by your great project.

As I dig deeper back into it now then, I am now just noticing one further quirk with your OpenCL Wrapper I do not fully understand how to work around and would appreciate your consideration on it.

## CUDA Synchronization

In CUDA, one can run multiple operations on a stream, which will create a queue in which those operations occur in. One can then run `cudaStreamSynchronize(stream)` to wait until that stream completes. Running the kernels in CUDA is always non-blocking for the thread, and you either wait for the stream to complete with this function, or for the device to finish everything with `cudaDeviceSynchronize()`.

This is a good system because it allows you to start multiple jobs in parallel on the GPU (I presume most can schedule and run parallel "streams" or Kernel queues, right?), and either wait for one or all to complete after you have started the various jobs in parallel and/or series.

## 1) OPEN CL WRAPPER ENQUEUE
By contrast, in the OpenCL wrapper, the queuing function appears to be just creating one single global queue for the device. We do not have any concept of "streams" or parallelization of the jobs from what I can tell. We can enqueue jobs via each Kernel we create, but these go into a global device queue which is processed sequentially.

This can be demonstrated by the following experiment. 

If you create a code demo like:

```
kernel void simple_add(global float* C, float adder) { 
	const uint n = get_global_id(0);
	C[n] = C[n] + adder;
}
kernel void simple_divide(global float* C, float divider) { 
	const uint n = get_global_id(0);
	C[n] = C[n] / divider;
}
kernel void do_nothing_kernel(global float* C) {
}
```

```
int main() {
	Device device(select_device_with_most_flops());

	const uint N = 1024u;
	Memory<float> C(device, N);

	for(uint n = 0u; n < N; n++) {
		C[n] = 1.0f; //start with 1
	}
	C.write_to_device();

	print_info("Value before kernel execution: C[0] = " + cl_to_string(C[0]));

	//define kernels:
	const float amtToAdd = 1;
	const float amtToDivide = 2;
	
	Kernel add_kernel(device, N, "simple_add", C, amtToAdd);
	Kernel divide_kernel(device, N, "simple_divide", C, amtToDivide);
	Kernel do_nothing_kernel(device, N, "do_nothing_kernel", C);
	
	//===============================================
	//LOOP TEST 
	//===============================================
	int numLoops = 1024;
	for (int i = 0; i < numLoops; i++) {
		add_kernel.enqueue_run(); //add 1
		divide_kernel.enqueue_run(); //divide by 2
	}
	do_nothing_kernel.finish_queue();
		
	C.read_from_device(); 

	//print result
	print_info("Value after kernel execution: C[0] = " + cl_to_string(C[0]) + " C[] " + cl_to_string(C[5]));
	
	wait();
	return 0;
}
```
This just runs a loop of:
```
C = 1;
C +=1;
C = C/2;
C +=1;
C = C/2;
```
We can see several things:

- The result is always 1.0 at the end, indicating it ran in perfect serial order throughout.
- `enqueue_run` and `finish_queue` appear to work globally on a device queue basis - this code runs `finish_queue` on a Kernel that was not even used, but was created for the same `device`, and thus it works.

If `enqueue_run` and `finish_queue` are running on a serial global device queue, does this not prevent the GPU from running multiple Kernels in parallel? Or what?

It seems this enqueuing and finishing would highly benefit from a concept like "streams". If we had `enqueue_run(stream)` and `finish_queue(stream)` this would allow similar parallelization and `cudaStreamSynchronize(stream)` synchronization as CUDA. Could such a thing be possible in OpenCL or with your system?

## 2) OPENCL WRAPPER RUN

If we do not queue, but rather just `run()` we are blocked until that job completes. And thus we must again then operate only in serial for that device. This `run()` does not allow parallel GPU operations.

## IDEA?

The basic structural issue I think (if I am not missing something) is that in `opencl.hpp` there is only one `cl::CommandQueue cl_queue` per `Device`.

If we could instead create "streams" via creating more `cl::CommandQueue` objects inside a Device which map to different device streams or GPU parallel processors, and maybe have a get function to tell us how many we can create (I am not sure how this works), then perhaps that could solve it? 

Streams could be created like `StreamCL stream (device, "name");` or just `StreamCL stream (device);`  which would add a `CommandQueue` inside that `Device` by that identifier.

In theory then we could at least have Kernel parallelization like this:

```
Device device(select_device_with_most_flops());
StreamCL stream1(device, "stream1");
StreamCL stream2(device, "stream2");

Kernel add_kernel(device, N, "simple_add", B, amtToAdd);
Kernel divide_kernel(device, N, "simple_divide", C, amtToDivide);

add_kernel.enqueue_run(stream1); 
add_kernel.enqueue_run(stream2); 

add_kernel.finish_queue(); //to run the two jobs in parallel on GPU as they are now on different device queues, wait until all are done
```

Conceptually it might I think also make more sense logically (given that the queues are Device level rather than Kernel level) to have this type of syntax but this is just cosmetic and I am not sure:

```
Device device(select_device_with_most_flops());
StreamCL stream1(device, "stream1");
StreamCL stream2(device, "stream2");

//Possibility 1: Perhaps optionally assign kernels to stream at creation:
Kernel add_kernel(device, N, "simple_add", B, amtToAdd, stream1); //optional stream argument for kernel to be assigned to
Kernel divide_kernel(device, N, "simple_divide", C, amtToDivide, stream2);  //optional stream argument

//Possibility 2: Run the functions from the Device class instead:
device.enqueue_run(add_kernel, stream1);  //optional stream queue argument for device to consider
device.enqueue_run(add_kernel, stream2);  //optional stream queue argument for device to consider

device.finish_queue(); //to run the two jobs in parallel on GPU as they are now on different device queues and wait until all are done
//device.finish_queue(stream1); //or state stream for device to just run and wait for that one stream queue to finish.
```

According to [this ](https://stackoverflow.com/questions/27240746/running-parallel-opencl-kernels) I think I am understanding the issue correctly at least though I could not fix it myself:

>The way to get kernels to run in parallel using OpenCL software only is to queue them with different command queues on the same device. 

What do you think? Am I missing something obvious? Or is this worth considering? Any other ideas? Thanks again.

**Comment by ProjectPhysX** [comment]
Hi @jonmdev,

most GPUs execute one kernel after another, even if they are enqueued in the same queue without finish_queue in between, or in entirely different queues. And even if each kernel alone only uses a single SM.

A few GPUs can do this, but most can't, and there is no guarantee of parallel asynchronous execution of multiple kernels.

To try it with multiple queues, create two Devices for the same GPU; they each have their own queue.

But if your application only dispatches a single workgroup/warp, I'd say it is entirely unsuited for GPU, and just doing it on the CPU with SIMD is way faster. The large PCIe communication latency eliminates any potential gains from parallelization. You need at least a couple 100k threads to fully saturate the GPU and get throughput benefits.

Kind regards,
Moritz

**Comment by jonmdev** [comment]
Thanks for the clarification. I am learning rapidly. I now have a working implementation at least which is great. 

I have learned how the concept of "work groups" is applied in OpenCL and how to query the max size for my device and kernels, as well as specify them, so I can adequately maximize capacity.

It is very fun now as the painful part is past, and I again appreciate your project. :)

Regarding the queuing and making multiple devices to get different queues, I tried that before posting but it didn't work, so I assumed the Device must be getting a static or instance reference (one of each "device"). 

But your explanation makes sense - probably my device did not support parallel operations so that is why it queued in serial. I will look at this more in the future but thanks for the help and instruction. It is nice of you to keep supporting this project.
